app/controllers/oaisys/pmh_controller.rb

```diff
@@ -17,59 +17,73 @@ class Oaisys::PMHController < Oaisys::ApplicationController
   def bad_verb; end
 
   def identify
-    expect_no_args for_verb: :Identify
-
     respond_to do |format|
       format.xml { render :identify }
     end
   end
 
   def list_sets
-    expect_args for_verb: :ListSets, exclusive: [:resumptionToken]
+    expect_args exclusive: [:resumptionToken]
   end
 
   # TODO: Handle the identifier argument.
   def list_metadata_formats
-    expect_args for_verb: :ListMetadataFormats, optional: [:identifier]
+    expect_args optional: [:identifier]
 
+    parameters = params.except('controller', 'action')
     respond_to do |format|
       format.xml do
-        render :list_metadata_formats, locals: { supported_formats: SUPPORTED_FORMATS }
+        render :list_metadata_formats, locals: { supported_formats: SUPPORTED_FORMATS, parameters: parameters }
       end
     end
   end
 
   def list_records
-    expect_args for_verb: :ListRecords, required: [:metadataPrefix], optional: [:from, :until, :set],
-                exclusive: [:resumptionToken]
+    expect_args required: [:metadataPrefix], optional: [:from, :until, :set], exclusive: [:resumptionToken]
   end
 
   # get_record is referring to the verb, not a getter.
   # rubocop:disable Naming/AccessorMethodName
   def get_record
-    expect_args for_verb: :GetRecord, required: [:identifier, :metadataPrefix]
+    expect_args required: [:identifier, :metadataPrefix]
   end
   # rubocop:enable Naming/AccessorMethodName
 
+  # TODO: Handle from, until, and resumptionToken arguments.
   def list_identifiers
-    expect_args for_verb: :ListIdentifiers, required: [:metadataPrefix], optional: [:from, :until, :set],
-                exclusive: [:resumptionToken]
+    expect_args required: [:metadataPrefix], optional: [:from, :until, :set], exclusive: [:resumptionToken]
+
+    parameters = params.except('controller', 'action').to_unsafe_h
+    metadata_prefix = params[:metadataPrefix]
+    if metadata_prefix == 'oai_dc'
+      results = Oaisys::Engine.config.oai_dc_model.where(visibility: JupiterCore::VISIBILITY_PUBLIC)
+    elsif metadata_prefix == 'oai_etdms'
+      results = Oaisys::Engine.config.oai_etdms_model.where(visibility: JupiterCore::VISIBILITY_PUBLIC)
+    else
+      parameters = params.to_unsafe_h.slice(:verb, :metadataPrefix)
+      raise Oaisys::CannotDisseminateFormatError.new(parameters: parameters)
+    end
+
+    results = results.where("member_of_paths": params[:set].tr(':', '/')) if params[:set].present?
```

Comment:
I'd turn this into a scope in the relevant models in Jupiter, so that we're working with models at arms length rather than making assumptions here that leave ourselves dependent on a specific kind of table structure.